python_modules/dagster/dagster/daemon/auto_run_reexecution/auto_run_reexecution.py

```diff
@@ -0,0 +1,176 @@
+
+import structlog
+
+from dagster import DagsterRun, MetadataEntry, MetadataValue
+from dagster.core.events import EngineEventData
+from dagster.core.execution.plan.resume_retry import ReexecutionStrategy
+from dagster.core.instance import DagsterInstance
+from dagster.core.storage.pipeline_run import DagsterRunStatus, RunRecord
+from dagster.core.storage.tags import MAX_RETRIES_TAG, RETRY_NUMBER_TAG
+from dagster.core.workspace.workspace import IWorkspace
+from dagster.utils.error import serializable_error_info_from_exc_info
+
+logger = structlog.get_logger("automatic_reexecution_consumer")
+
+DEFAULT_MAX_RETRIES = 0
+
+
+def filter_runs_to_should_retry(
+    runs: List[DagsterRun], instance: DagsterInstance, default_max_retries: int
+) -> Iterator[Tuple[DagsterRun, int]]:
+    """
+    Return only runs that should retry along with their retry number (1st retry, 2nd, etc.)
+    """
+
+    def get_retry_number(run: DagsterRun) -> Optional[int]:
+        if run.status != DagsterRunStatus.FAILURE:
+            return None
+
+        raw_max_retries_tag = run.tags.get(MAX_RETRIES_TAG)
+        if raw_max_retries_tag is None:
+            max_retries = default_max_retries
+        else:
+            try:
+                max_retries = int(raw_max_retries_tag)
+            except ValueError:
+                instance.report_engine_event(
+                    f"Error parsing int from tag {MAX_RETRIES_TAG}, won't retry the run.", run
+                )
+                return None
+
+        if max_retries == 0:
+            return None
+
+        # NOTE: group these to reduce db calls
+        run_group = instance.get_run_group(run.run_id)
+
+        if run_group:
+            _, run_group_list = run_group
+
+            # Has the parent run already been retried the maximum number of times? (Group includes the parent)
+            if len(run_group_list) >= max_retries + 1:
+                return None
+
+            # Does this run already have a child run?
+            if any([run.run_id == run_.parent_run_id for run_ in run_group_list]):
+                return None
+
+        return 1 if not run_group else len(run_group_list)
+
+    for run in runs:
+        retry_number = get_retry_number(run)
+        if retry_number is not None:
+            yield (run, retry_number)
+
+
+def retry_run(
+    failed_run: DagsterRun,
+    retry_number: int,
+    instance: DagsterInstance,
+    workspace: IWorkspace,
+) -> None:
+    """
+    Submit a retry as a re-execute from failure
+    """
+
+    tags = {RETRY_NUMBER_TAG: str(retry_number)}
+
+    if not failed_run.external_pipeline_origin:
+        instance.report_engine_event(
+            "Run does not have an external pipeline origin, unable to retry the run.",
+            failed_run,
+        )
+        logger.error(
+            "Run does not have an external pipeline origin, this shouldn't be possible unless users are creating runs manually?",
```

Comment:
The copy here is a little odd for a user-facing error, maybe drop the question mark and make it 2nd-person instead of third-person